File: ackm/Лабораторна робота 2/lab2.m
disp("Інтегрування:")
quad('sqrt(4 - x**2)/x',0.2,1,0.01)
disp("Диференціювання:")
x_start = 0.2;
x_end = 1;
h_dif = 0.01;
N = 40;

f = @(x) sqrt(4 - x.^2);
F = @(x) sqrt(4 - x.^2) - 2*log((2 + sqrt(4 - x.^2)) / x);

x = linspace(x_start, x_end, N);

h = (x_end - x_start) / 2 * (N - 1);
xi = x_start:h:x_end;

real = h / 3 * (f(xi(1)) + 2 * sum(f(xi(3: 2: end - 2))) + 4 * sum(f(xi(2: 2: end))) + f(xi(end)));

real_result = F(x_end) - F(x_start);
real;


x = 0.5;
x0 = x_start:h_dif:x_end;

n = size(x0, 2) - 1;
new_real = 0;

for idx=1:n
    if ((x < x0(idx+1)) && (x >= x0(idx)))
        if (idx == 0)
            new_real += (-3 * F(x0(idx)) + 4 * F(x0(idx + 1)) - F(x0(idx + 2))) / (2 * h_dif);
            break;
        elseif (idx == n)
            new_real += (3 * F(x0(idx)) - 4 * F(x0(idx - 1)) + F(x0(idx - 2))) / (2 * h_dif);
            break;
        else
            new_real += (F(x0(idx + 1)) - F(x0(idx - 1))) / (2 * h_dif);
            break;
        endif;
    endif;
endfor;

disp(new_real);

File: ackm/Лабораторна робота 2/lab2.py
import math

N = 40
h = 0.01
x_data = [0.2, 1]
f = lambda x: math.sqrt(4 - x ** 2) / x
F = lambda x: math.sqrt(4 - x ** 2) - 2 * math.log((2 + math.sqrt(4 - x ** 2)) / x)

newton_leibniz = lambda B, A, func: func(B) - func(A)


def f_by_step(x_start, func, step, n):
    return [func(x_start + idx * step) for idx in range(n + 1)]


print("Дифференціювання методом многочленів Лагранжа")


def lagrange_polinom(f_arr, i, step, n):
    if i == 0:
        return (-3 * f_arr[i] + 4 * f_arr[i + 1] - f_arr[i + 2]) / (2 * step)
    elif i == n:
        return (f_arr[i - 2] - 4 * f_arr[i - 1] + 3 * f_arr[i]) / (2 * step)
    else:
        return (f_arr[i + 1] - f_arr[i - 1]) / (2 * step)


def lagrange(func, interval, step, x):
    x_start, x_end = interval
    n = int((x_end - x_start) / step)
    f_arr = f_by_step(x_start, func, step, n)

    for idx in range(n):
        if x_start + idx * step <= x and x < x_start + (idx + 1) * step:
            return lagrange_polinom(f_arr, idx, step, n)


x_ = 0.5
result = lagrange(F, x_data, h, x_)
real_result = f(x_)
print(f"Iстинне значення :  {real_result}; Отримане значення:  {result}")
print(
    f"Абсолютна похибка:  {abs(result-real_result)};  Вiдносна похибка: {abs(result-real_result) / real_result * 100} %"
)

print("Інтегрування методом Сімпсона")


def simpson(func, interval, n):
    x_start, x_end = interval

    step = (x_end - x_start) / n

    # Odd
    summ_e_o = 0
    x = x_start + step
    for i in range(n // 2):
        summ_e_o += 4 * func(x)
        x += 2 * step

    # Even
    x = x_start + 2 * step
    for j in range(n // 2 - 1):
        summ_e_o += 2 * func(x)
        x += 2 * step

    return step * (func(x_start) + func(x_end) + summ_e_o) / 3


result = simpson(f, x_data, 2 * N)
real_result = newton_leibniz(*x_data[::-1], F)
print(
    f"Iстинне значення iнтегралу:  {real_result};  Наближене за {N} промiжкiв значення:  {result}"
)
print(
    f"Абсолютна похибка:  {abs(result-real_result)};  Вiдносна похибка: {abs(result-real_result) / real_result * 100} %"
)
